File: internal/utils/crypto.go
package utils

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"crypto/subtle"
	"encoding/base64"
	"errors"
	"io"
	"runtime"
	"strconv"
	"strings"

	"golang.org/x/crypto/argon2"
	"golang.org/x/crypto/bcrypt"
)

// Argon2id parameters - tuned defaults for server-side password hashing.
// Memory is expressed in KB for golang.org/x/crypto/argon2.
const (
	argonTime   = 3     // number of iterations
	argonMemory = 65536 // 64 MB (in KB)
	argonKeyLen = 32
	saltLen     = 16
)

func HashPassword(password string) (string, error) {
	if password == "" {
		return "", errors.New("password required")
	}

	// generate random salt
	salt := make([]byte, saltLen)
	if _, err := rand.Read(salt); err != nil {
		return "", err
	}

	threads := uint8(runtime.NumCPU())

	// derive key
	hash := argon2.IDKey([]byte(password), salt, uint32(argonTime), uint32(argonMemory), threads, uint32(argonKeyLen))

	b64Salt := base64.StdEncoding.EncodeToString(salt)
	b64Hash := base64.StdEncoding.EncodeToString(hash)

	encoded := strings.Join([]string{
		"$argon2id",
		"v=19",
		"m=" + strconv.Itoa(argonMemory) + ",t=" + strconv.Itoa(argonTime) + ",p=" + strconv.Itoa(int(threads)),
		b64Salt,
		b64Hash,
	}, "$")

	return encoded, nil
}

func CheckPasswordHash(password, encodedHash string) bool {
	if password == "" || encodedHash == "" {
		return false
	}

	// backward compatibility: bcrypt hashes start with $2a$ or $2b$ or $2y$
	if strings.HasPrefix(encodedHash, "$2a$") || strings.HasPrefix(encodedHash, "$2b$") || strings.HasPrefix(encodedHash, "$2y$") {
		// verify with bcrypt
		if bcrypt.CompareHashAndPassword([]byte(encodedHash), []byte(password)) == nil {
			return true
		}
		return false
	}

	parts := strings.Split(encodedHash, "$")
	// expected parts: "", "argon2id", "v=19", "m=...,t=...,p=...", "salt", "hash"
	if len(parts) != 6 || parts[1] != "argon2id" {
		return false
	}

	paramsPart := parts[3]
	saltB64 := parts[4]
	hashB64 := parts[5]

	// parse params
	var mem, t, p int
	for _, kv := range strings.Split(paramsPart, ",") {
		if strings.HasPrefix(kv, "m=") {
			mem, _ = strconv.Atoi(strings.TrimPrefix(kv, "m="))
		} else if strings.HasPrefix(kv, "t=") {
			t, _ = strconv.Atoi(strings.TrimPrefix(kv, "t="))
		} else if strings.HasPrefix(kv, "p=") {
			p, _ = strconv.Atoi(strings.TrimPrefix(kv, "p="))
		}
	}

	salt, err := base64.StdEncoding.DecodeString(saltB64)
	if err != nil {
		return false
	}

	expectedHash, err := base64.StdEncoding.DecodeString(hashB64)
	if err != nil {
		return false
	}

	computed := argon2.IDKey([]byte(password), salt, uint32(t), uint32(mem), uint8(p), uint32(len(expectedHash)))

	if subtle.ConstantTimeCompare(computed, expectedHash) == 1 {
		return true
	}

	return false
}

// Encrypt encrypts plaintext with a password-derived key using Argon2id (per-encryption salt) and AES-GCM.
// Output format (text): "v1$<salt_b64>$<nonce_b64>$<ct_b64>"
func Encrypt(plaintext, password string) (string, error) {
	if password == "" {
		return "", errors.New("password required")
	}

	// per-encryption random salt
	salt := make([]byte, saltLen)
	if _, err := io.ReadFull(rand.Reader, salt); err != nil {
		return "", err
	}

	threads := uint8(runtime.NumCPU())
	key := argon2.IDKey([]byte(password), salt, uint32(argonTime), uint32(argonMemory), threads, uint32(argonKeyLen))

	block, err := aes.NewCipher(key)
	if err != nil {
		return "", err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	nonce := make([]byte, gcm.NonceSize())
	if _, err := io.ReadFull(rand.Reader, nonce); err != nil {
		return "", err
	}

	ciphertext := gcm.Seal(nil, nonce, []byte(plaintext), nil)

	bSalt := base64.StdEncoding.EncodeToString(salt)
	bNonce := base64.StdEncoding.EncodeToString(nonce)
	bCT := base64.StdEncoding.EncodeToString(ciphertext)

	encoded := strings.Join([]string{"v1", bSalt, bNonce, bCT}, "$")
	return encoded, nil
}

// Decrypt reads the versioned format produced by Encrypt and attempts to decrypt using Argon2id-derived key.
func Decrypt(ciphertextStr, password string) (string, error) {
	if password == "" || ciphertextStr == "" {
		return "", errors.New("password and ciphertext required")
	}

	parts := strings.Split(ciphertextStr, "$")
	// expected: "v1" "salt" "nonce" "ct"
	if len(parts) != 4 || parts[0] != "v1" {
		return "", errors.New("unsupported ciphertext format")
	}

	salt, err := base64.StdEncoding.DecodeString(parts[1])
	if err != nil {
		return "", err
	}
	nonce, err := base64.StdEncoding.DecodeString(parts[2])
	if err != nil {
		return "", err
	}
	ct, err := base64.StdEncoding.DecodeString(parts[3])
	if err != nil {
		return "", err
	}

	threads := uint8(runtime.NumCPU())
	key := argon2.IDKey([]byte(password), salt, uint32(argonTime), uint32(argonMemory), threads, uint32(argonKeyLen))

	block, err := aes.NewCipher(key[:])
	if err != nil {
		return "", err
	}

	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	if len(nonce) != gcm.NonceSize() {
		return "", errors.New("malformed ciphertext (nonce size mismatch)")
	}

	plaintext, err := gcm.Open(nil, nonce, ct, nil)
	if err != nil {
		return "", err
	}

	return string(plaintext), nil
}

// HashAPIKey creates a keyed HMAC-SHA256 of an API token. Store only the returned value in the database.
// The hmacKey should be a server-managed secret (pepper) kept out of the repository/DB; pass it from config.
func HashAPIKey(token, hmacKey string) (string, error) {
	if token == "" || hmacKey == "" {
		return "", errors.New("token and hmac key required")
	}
	mac := hmac.New(sha256.New, []byte(hmacKey))
	if _, err := mac.Write([]byte(token)); err != nil {
		return "", err
	}
	sum := mac.Sum(nil)
	return base64.StdEncoding.EncodeToString(sum), nil
}

// CompareAPIKey verifies a token against a stored HMAC-SHA256 (base64 encoded).
func CompareAPIKey(token, storedB64, hmacKey string) bool {
	if token == "" || storedB64 == "" || hmacKey == "" {
		return false
	}
	stored, err := base64.StdEncoding.DecodeString(storedB64)
	if err != nil {
		return false
	}
	mac := hmac.New(sha256.New, []byte(hmacKey))
	if _, err := mac.Write([]byte(token)); err != nil {
		return false
	}
	expected := mac.Sum(nil)
	return subtle.ConstantTimeCompare(stored, expected) == 1
}

File: src/utils/crypto.ts
/*
 * resure — the best way to store your secrets instead of forgetting them.
 *
 * Copyright (c) 2025-present
 * Shubham Yadav (bas3line), Nandha (almightynan)
 * License: MIT
 *
 * Permission is hereby granted, free of charge, to any person obtaining a
 * copy of this software and associated documentation files (the "Software"),
 * to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 */

// this file contains crypto helpers used across the server.
// all comments added by the assistant are lowercase as requested.
// responsibilities:
// - password hashing and verification
// - api key generation and hmac hashing
// - per-secret encryption and decryption using argon2id + xchacha20-poly1305
// notes:
// - ciphertext format is: v1$<salt_b64>$<nonce_b64>$<ct_b64>
// - argon2 parameters are chosen for moderate hardness; tune in production

import crypto from 'crypto'
import argon2 from 'argon2'
import { XChaCha20Poly1305 } from '@stablelib/xchacha20poly1305'

const ARGON_TIME = 3
const ARGON_MEMORY = 65536
const ARGON_KEYLEN = 32
const SALT_LEN = 16
const NONCE_LEN = 24 // XChaCha20-Poly1305 nonce size

// hashPassword
// - wrapper around argon2.hash
// - returns a phc encoded argon2 string suitable for storage in users.password
// - keep this async so callers can await and surface errors cleanly
export async function hashPassword(password: string) {
  return await argon2.hash(password)
}

// verifyPassword
// - verify a plain password against a stored argon2 hash
// - returns true on match, false on any error or mismatch
export async function verifyPassword(hash: string, password: string) {
  try {
    return await argon2.verify(hash, password)
  } catch (err) {
    // verification failure or malformed hash -> treat as non-match
    return false
  }
}

// hashAPIKey
// - deterministic hmac-sha256 over a raw api key using a pepper value
// - the pepper must be secret and rotated carefully
// - stored value is hex digest
export function hashAPIKey(raw: string, pepper: string) {
  const hmac = crypto.createHmac('sha256', pepper)
  hmac.update(raw)
  return hmac.digest('hex')
}

// generateRawAPIKey
// - produce a human safe-ish prefix then 32 random bytes hex
// - callers should show the raw key once and persist only the hmac in db
export function generateRawAPIKey() {
  return 'svp_' + crypto.randomBytes(32).toString('hex')
}

// encrypt
// - produces ciphertext in versioned format the go backend expects: v1$salt$nonce$ciphertext
// - workflow:
//   1) generate random salt
//   2) derive a 32 byte key using argon2id with the salt
//   3) encrypt with xchacha20-poly1305 using a random 24 byte nonce
//   4) return base64 parts joined with dollar separators
// - note: callers must provide the password (master password) used to derive the key
export async function Encrypt(plaintext: string, password: string) {
  if (!password) throw new Error('password required')
  const salt = crypto.randomBytes(SALT_LEN)
  const threads = Math.max(1, require('os').cpus().length)

  // derive raw key via argon2id (returns Buffer when raw: true)
  const key = (await argon2.hash(password, {
    type: (argon2 as any).argon2id,
    salt,
    timeCost: ARGON_TIME,
    memoryCost: ARGON_MEMORY,
    parallelism: threads,
    hashLength: ARGON_KEYLEN,
    raw: true,
  } as any)) as Buffer

  const nonce = crypto.randomBytes(NONCE_LEN)
  const aead = new XChaCha20Poly1305(key)
  const ciphertext = aead.seal(nonce, Buffer.from(plaintext, 'utf8'))

  const bSalt = salt.toString('base64')
  const bNonce = nonce.toString('base64')
  const bCT = Buffer.from(ciphertext).toString('base64')

  return `v1$${bSalt}$${bNonce}$${bCT}`
}

// decrypt
// - accepts ciphertext in the v1 format and the password used to derive the key
// - derives key with the same argon2 parameters and attempts open()
// - throws on malformed input or failed authentication
export async function Decrypt(ciphertextStr: string, password: string) {
  if (!password || !ciphertextStr) throw new Error('password and ciphertext required')
  if (!ciphertextStr.startsWith('v1$')) {
    throw new Error('unsupported ciphertext format')
  }
  const parts = ciphertextStr.split('$')
  if (parts.length !== 4) throw new Error('malformed ciphertext')
  const salt = Buffer.from(parts[1], 'base64')
  const nonce = Buffer.from(parts[2], 'base64')
  const ct = Buffer.from(parts[3], 'base64')

  const threads = Math.max(1, require('os').cpus().length)
  const key = (await argon2.hash(password, {
    type: (argon2 as any).argon2id,
    salt,
    timeCost: ARGON_TIME,
    memoryCost: ARGON_MEMORY,
    parallelism: threads,
    hashLength: ARGON_KEYLEN,
    raw: true,
  } as any)) as Buffer // reminds me of my friend bufferwise :3

  const aead = new XChaCha20Poly1305(key)
  const plaintext = aead.open(nonce, ct)
  if (!plaintext) throw new Error('decryption failed')
  return Buffer.from(plaintext).toString('utf8')
}
